[TEP-0092] TaskRun scheduling timeout ⏱️

bobcatfish opened the pull request:
This TEP addresses https://github.com/tektoncd/pipeline/issues/4078 by
proposing that at runtime `TaskRuns` can be configured with a new
timeout that applies before they actually start executing - and only
after that are the existing timeouts applied.

This is meant to make it possible to control when our existing timeout
feature works in resource constrained environment and also to get parity
with [Google Cloud Build's existing queuettl feature](https://cloud.google.com/build/docs/build-config-file-schema#queuettl).

pritidesai commented:
API WG - @bobcatfish responding to comments

bobcatfish commented:
/remove-lifecycle rotten

I am going to update this 🙏

badamowicz commented:
As the original author of this [issue](https://github.com/tektoncd/pipeline/issues/4078) may I ask for which Tekton version this TEP is scheduled? Would help us in our planning. Thanks!

bobcatfish commented on teps/0092-scheduling-timeout.md:
```diff
@@ -0,0 +1,363 @@
+  This would mean the timeout would also include the time for init containers to execute (which Pipelines still uses
+  for some features) as well as the time to pull any images needed by the containers. `Ready` would also include a
+  positive indication from any readiness probes that might have been added to the pod.
+* We could stop the timeout once [the Pod phase is `Running`](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#Pod-phase),
```
(discussing in comment above https://github.com/tektoncd/community/pull/540#discussion_r735661039)

bobcatfish commented on teps/0092-scheduling-timeout.md:
```diff
@@ -0,0 +1,363 @@
+been added to the pod, indicating that a node is available for the pod to run on and it has been assigned to that node.
+
+Assumptions:
+* We don't want to include init container execution time in the timeout
+* We don't want to include image pull time in the timeout
```
@imjasonh would you be okay with proceeding without the image pull timeout or the init container execution being included in the scheduling timeout, and then potentially adding these timeouts as an additional feature if needed later?

Overall my reason for preferring this is that image pull time and time to execute init contains feel like they address separate use cases than what we're trying to address here

The situations a user might want to address here (i.e. image pull backoff, long image pulls and long init container execution) seem like they're all fairly distinct and the kinds of timeouts users would want to provide would be very different, so bundling them into one option doesn't totally make sense to me.

At the very least I'm adding a section about this in the alternatives and adding an example of how we could still address it in the future.

bobcatfish commented on teps/0092-scheduling-timeout.md:
````diff
@@ -0,0 +1,363 @@
+
+```yaml
+spec:
+  timeouts:
+    pipeline: "0h0m60s"
+    tasks: "0h0m40s"
+    taskrun-scheduling: "5m" # this is the new option
+    finally: "0h0m20s"
````
i think this is accurate - a few lines below this is a description of how the timeout logic would apply, and like you said, the scheduling timeout would be constrained by the pipeline timeout

the example times don't make any sense b/c the scheduling timeout is 5 minutes and the pipeline only has 60 seconds total to execute, ill fix that 😆 

I can see why you might want the scheduling timeout to be completely separate - if you want to allow your pipeline to run for 5 min total regardless of how long it takes to schedule each individual task, but I can't really think of a better solution and I think it's probably more common to want to constrain the overall time for a pipeline vs to want to say "overall time except for the scheduling time" @badamowicz let me know if you disagree - i.e. if it's important for your use case to be able to say "I want the pipeline execution to take only 5 minutes, even if it takes 30 minutes per pod to schedule"

Another thing to consider, if you only cared about the scheduling timeout and you didn't care about the overall execution time, you could always let those top level timeouts be infinite (I'm pretty sure 0 means no timeout, right? im a bit confused by our docs on this, will follow up with @jerop ), e.g. something like:

```yaml
spec:
  timeouts:
    pipeline: "0s"
    taskruns:
      scheduling: "1m" # this is the new option
```

bobcatfish commented:
@jerop:

> Made me wonder about taskrun-execution which sets the execution timeout for each task, but we don't have use cases for this and can be explored later if needed

now that i've updated the syntax to allow for more task level timeouts in future I think we could account for this if we had use cases in the future, e.g.:

```yaml
spec:
  timeouts:
    pipeline: "0h15m0s"
    tasks: "0h10m0s"
    finally: "0h5m0s"
    taskruns: # new section
      scheduling: "1m" # new option
      execution: "1m" # could be added in future if needed
```

> Noting here that we discussed in the API WG on 10/18 that scheduling concerns have come up in other areas and we could explore solving them together

I can see the overlap in these concerns all being scheduling related (maybe a scheduling "problem space"? :D) but addressing parallel execution and coordination between the scheduling of tasks feels to me like something we'd address separately - my reasoning is that in this TEP we're not changing the scheduling behavior, we're just saying we want to stop if things are happening too slowly, vs. those issues seem to be about actually changing what is scheduled and when? (not sure how convincing that is!)

> Could providing Pending TaskRuns, where the existing timeouts don't apply while they're pending, be sufficient for the use cases discussed here? Maybe we could add this as an alternative approach to solve for the use cases in this TEP? 

I'll add this as an alternative ("Leverage pending status") - let me know if I didn't understand the suggestion, my assumption is that in that solution users would need to build their own controller to manage pending TaskRuns and apply the timeouts logic they want.

(There's also another alternative listed in the bullet list of alternatives that involves Tekton Pipelines moving all timeout logic into a separate more customizable controller.)

bobcatfish commented:
@imjasonh @jerop @afrittoli @souleb after a significant delay I've finally responded to your feedback and updated this TEP - I've also ambitiously updated it to "implementable" 🙏  PTAL!

ive made a few updates:
- implementable (instead of just proposed)
- upated the syntax to create a new "taskruns" section within the pipelinerun level configuration to:
  - a. make it more clear that this option applies per task(run)
  - b. make it possible to add more task(run) level timeouts in future if needed
- tried to make the example of how the timeouts interact with each other more clear (values didnt make sense before)
- Added alternative: including image pull time in scheduling timeout; also highlighting that we could add image pull time specific timeouts later (let me know what you think @imjasonh )
- Added alternative: leveraging "pending" status

@badamowicz my apologies for this extreme delay 😬  !! to answer your question:

> As the original author of this https://github.com/tektoncd/pipeline/issues/4078 may I ask for which Tekton version this TEP is scheduled? Would help us in our planning. Thanks!

At the moment this isn't explicitly scheduled into a particular Tekton version. Implementing this particular feature is unfortunately not high priority for the Google team (so if anyone reading this is interested in volunteering to do the implementation that is very welcome!!) BUT we have some new folks starting on the team soon so my plan is to get this proposal marked as implementable so we can have one of our new team members take it on as part of their ramp up.

bobcatfish commented:
hahaha i said PTAL @afrittoli b/c i was convinced you left feedback on this but apparently i imagined it... XD

badamowicz commented:
@bobcatfish Thanks for the update! Meanwhile I've left the project this issue occurred. However, will inform my colleagues to further track this issue.

imjasonh commented:
> Added alternative: including image pull time in scheduling timeout; also highlighting that we could add image pull time specific timeouts later (let me know what you think @imjasonh )

This seems right to me. 👍 

/lgtm

bobcatfish commented:
/hold

jerop reviewed (approved):
thanks @bobcatfish!

dfuessl commented:
Hi @bobcatfish,
thank you for all of your activities in this ticket so far. I am taking over from @badamowicz because he has left the project. I was involved in our internal discussion and I know all the details.

bobcatfish commented:
Welcome @dfuessl !! :D 🎉 🎉 🎉   and thanks again for your team's patience around this

bobcatfish commented:
/hold cancel

dibyom commented:
/lgtm
